Prox/Prox/Models/PlaceProviders.swift

```diff
@@ -0,0 +1,159 @@
+        photoURLs = photoURLStrings.flatMap { URL(string: $0) }
+
+        if let urlString = dict["url"] as? String,
+           let url = URL(string: urlString) {
+            self.url = url
+        } else {
+            self.url = nil
+        }
+
+        address = (dict["address"] as? [String])?.joined(separator: " ")
+
+        if let hoursDictFromServer = dict["hours"] as? [String : [[String]]],
+            let hoursFromServer = OpenHours.fromFirebaseValue(hoursDictFromServer) {
+            hours = hoursFromServer
+        } else {
+            hours = nil
+        }
+
+        rating = dict["rating"] as? Float
+
+        totalReviewCount = dict["totalReviewCount"] as? Int
+    }
+}
+
+/// Pulls data from multiple providers, prioritizing data in the order of the providers given.
+class CompositePlaceProvider: PlaceProvider {
+    private(set) var id: String?
+    private(set) var name: String?
+    private(set) var description: String?
+    private(set) var categories = (names: [String], ids: [String])([], [])
+    private(set) var latLong: CLLocationCoordinate2D?
+    private(set) var photoURLs = [URL]()
+    private(set) var url: URL?
+    private(set) var address: String?
+    private(set) var hours: OpenHours?
+    private(set) var rating: Float? = nil
+    private(set) var totalReviewCount: Int? = nil
+
+    init(fromProviders providers: [PlaceProvider]) {
+        for provider in providers {
+            if id == nil {
+                id = provider.id
+            }
+
+            if name == nil {
+                name = provider.name
+            }
+
+            if description == nil {
+                description = provider.description
+            }
+
+            // TODO: We could merge multiple categories if we wanted.
+            if self.categories.ids.isEmpty {
+                self.categories = provider.categories
+            }
+
+            if latLong == nil {
+                latLong = provider.latLong
+            }
+
+            // TODO: We could merge multiple photos if we wanted.
+            if photoURLs.isEmpty {
+                photoURLs = provider.photoURLs
+            }
+
+            // TODO: We could merge multiple URLs if we wanted.
+            if url == nil {
+                url = provider.url
+            }
+
+            if address == nil {
+                address = provider.address
+            }
+
+            if hours == nil {
+                hours = provider.hours
+            }
+        }
+    }
+}
+
+extension SourcePlaceProvider {
+    fileprivate static func getCategories(fromFirebaseValue value: [[String:String]]) -> (names: [String], ids: [String]) {
+        var names = [String]()
+        var ids = [String]()
+        for category in value {
+            guard let name = category["text"],
+                let id = category["id"] else {
+                    print("lol unable to retrieve category from firebase data for place")
+                    return ([], [])
```

Comment:
Maybe we should `continue` rather than `returning` early if we're missing just one name/category pair.